File: kotlin/day2_part2_2019.kt
fun main() {
    val input = "input.txt".readFile()
    val program = input.split(",").map { it.toInt() }.toMutableList()

    program[1] = 12
    program[2] = 2
    runProgram(program)
    println(program[0])

    val targetOutput = 19690720
    for (noun in 0..99) {
        for (verb in 0..99) {
            val memory = input.split(",").map { it.toInt() }.toMutableList()
            memory[1] = noun
            memory[2] = verb
            runProgram(memory)
            if (memory[0] == targetOutput) {
                println(100 * noun + verb)
                return
            }
        }
    }
}

fun runProgram(program: MutableList<Int>) {
    var ip = 0
    while (true) {
        when (program[ip]) {
            1 -> {
                val a = program[program[ip + 1]]
                val b = program[program[ip + 2]]
                program[program[ip + 3]] = a + b
            }
            2 -> {
                val a = program[program[ip + 1]]
                val b = program[program[ip + 2]]
                program[program[ip + 3]] = a * b
            }
            99 -> return
        }
        ip += 4
    }
}

fun String.readFile(): String {
    return java.io.File(this).readText().trim()
}

File: swift/day2_part2_2019.swift

import Foundation

let fileURL = URL(fileURLWithPath: "input.txt")
let input = try String(contentsOf: fileURL)
var program = input.split(separator: ",").map { Int($0)! }

func runProgram(noun: Int, verb: Int) -> Int {
    var memory = program
    memory[1] = noun
    memory[2] = verb

    var instructionPointer = 0

    while memory[instructionPointer] != 99 {
        let opcode = memory[instructionPointer]
        let input1 = memory[memory[instructionPointer + 1]]
        let input2 = memory[memory[instructionPointer + 2]]
        let outputPosition = memory[instructionPointer + 3]

        if opcode == 1 {
            memory[outputPosition] = input1 + input2
        } else if opcode == 2 {
            memory[outputPosition] = input1 * input2
        }

        instructionPointer += 4
    }

    return memory[0]
}

let targetOutput = 19690720
var result = 0

for noun in 0...99 {
    for verb in 0...99 {
        if runProgram(noun: noun, verb: verb) == targetOutput {
            result = 100 * noun + verb
            break
        }
    }
}

print(result)
